beacon_chain/validation/src/attestation_validation.rs

```diff
@@ -0,0 +1,256 @@
+};
+use super::message_generation::generate_signed_message;
+use super::signature_verification::{
+    verify_aggregate_signature_for_indices,
+    SignatureVerificationError,
+};
+
+#[derive(Debug,PartialEq)]
+pub enum AttestationValidationError {
+    ParentSlotTooHigh,
+    ParentSlotTooLow,
+    BlockSlotTooHigh,
+    BlockSlotTooLow,
+    JustifiedSlotIncorrect,
+    InvalidJustifiedBlockHash,
+    TooManyObliqueHashes,
+    BadCurrentHashes,
+    BadObliqueHashes,
+    BadAttesterMap,
+    IntWrapping,
+    PublicKeyCorrupt,
+    NoPublicKeyForValidator,
+    BadBitfieldLength,
+    InvalidBitfield,
+    InvalidBitfieldEndBits,
+    NoSignatures,
+    NonZeroTrailingBits,
+    BadAggregateSignature,
+    DBError(String),
+}
+
+/// The context against which some attestation should be validated.
+pub struct AttestationValidationContext<T>
+    where T: ClientDB + Sized
+{
+    /// The slot as determined by the system time.
+    pub block_slot: u64,
+    /// The slot of the parent of the block that contained this attestation.
+    pub parent_block_slot: u64,
+    /// The cycle_length as determined by the chain configuration.
+    pub cycle_length: u8,
+    /// The last justified slot as per the client's view of the canonical chain.
+    pub last_justified_slot: u64,
+    /// A vec of the hashes of the blocks preceeding the present slot.
+    pub parent_hashes: Arc<Vec<Hash256>>,
+    /// The store containing block information.
+    pub block_store: Arc<BlockStore<T>>,
+    /// The store containing validator information.
+    pub validator_store: Arc<ValidatorStore<T>>,
+    /// A map of (slot, shard_id) to the attestation set of validation indices.
+    pub attester_map: Arc<AttesterMap>,
+}
+
+impl<T> AttestationValidationContext<T>
+    where T: ClientDB
+{
+    /// Validate a (fully deserialized) AttestationRecord against this context.
+    ///
+    /// The function will return a HashSet of validator indices (canonical validator indices not
+    /// attestation indices) if the validation passed successfully, or an error otherwise.
+    ///
+    /// The attestation's aggregate signature will be verified, therefore the function must able to
+    /// access all required validation public keys via the `validator_store`.
+    pub fn validate_attestation(&self, a: &AttestationRecord)
+        -> Result<HashSet<usize>, AttestationValidationError>
+    {
+        /*
+         * The attesation slot must be less than or equal to the parent of the slot of the block
+         * that contained the attestation.
+         */
+        if a.slot > self.parent_block_slot {
+            return Err(AttestationValidationError::ParentSlotTooHigh);
+        }
+
+        /*
+         * The slot of this attestation must not be more than cycle_length + 1 distance
+         * from the parent_slot of block that contained it.
+         */
+        if a.slot < self.parent_block_slot
+            .saturating_sub(u64::from(self.cycle_length).saturating_add(1)) {
```

Comment:
I think the order of these "saturating" operations will give you a slight error wrt the spec.
"Verify that `slot >= max(parent.slot - CYCLE_LENGTH + 1, 0)`" -- the max operation here allows for `a.slot` to be 0, whereas your implementation does not.

For example say `parent_block_slot = 5, cycle_length = 10` then I believe the right side of your boolean is `1`, and prevents `a.slot` from equalling `0`.  If instead your logic matched the spec, the right side would be `0` and `a.slot` would not be less and would thus be valid.

That was a little verbose.. hope it makes sense.

Reply:
Unless stringing together saturating ops does something different than I expect.

Reply:
I had a look and I think what I have is correct. For reference, here's your parameters returning `0`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=90463a4a4df484611637787a97d021e8

Am I missing something?

Reply:
ah, interesting so the calls to `saturating` string together before resolving.

I was expecting `parent_block.saturating_sub(cycle_length)` to give 0 and then `0.saturating_add(1)` to give 1.

Sorry. Learning rust on the fly

Reply:
oh wait.
I missed the parens. Thought it was `parent_slot.saturating_sub(cycle_length).saturating_add(1)`

Reply:
Sneaky parens